---
title: "R Notebook"
output: html_notebook
---

Library Load
```{r}
library(ggplot2)
library(magrittr)
library(dplyr)
# install.packages("broom")
# install.packages("ggbeeswarm")
library(ggbeeswarm)
library(broom)
# install.packages("ggpubr")
library(ggpubr)
# install.packages("patchwork")
library(patchwork)
```

```{r}
hugo_epi_age = read.csv("/Users/hugohudson/Desktop/epigenetic_age_added.csv")
random_var <- hugo_epi_age$Epigenetic.Age

filtered_epi <- filter(hugo_epi_age,  Epigenetic.Age > 0)
epi_age_lm = lm(ageDeath ~ filtered_epi$Epigenetic.Age, data = filtered_epi)

epigenetic_ages <- data.frame(Epigenetic.Age = filtered_epi$Epigenetic.Age)
predicted_ages <- predict(epi_age_lm, newdata = epigenetic_ages)
model_residuals <- -residuals(epi_age_lm)

ggplot(data = filtered_epi, mapping = aes(x = ageDeath, y = predicted_ages, color = primaryDiagnosis)) +
  geom_point() + 
  geom_abline(slope = 1, intercept = 0) + 
  xlim(0, 100) + ylim(0, 100)

ggplot(data = filtered_epi, mapping = aes(x = ageDeath, y = model_residuals, color = primaryDiagnosis)) + 
  geom_point() + 
  geom_hline(yintercept = 0) +
  geom_abline(slope = 1, intercept = 0)

ggplot(data = filtered_epi, mapping = aes(x = primaryDiagnosis, y = model_residuals)) + 
  geom_quasirandom() +
  stat_compare_means()

epi_age_vs_sst_proportions <- ggplot(data = filtered_epi, mapping = aes(color = primaryDiagnosis, x = predicted_ages, y = SST)) +
  geom_point() + 
  labs( title = "Epigenetic Age vs SST Proportions", 
        x = "Epigenetic Age", 
        y = "SST Proportion") +
  stat_smooth(method = "lm") + 
  stat_cor() 

chrono_age_vs_sst_proportions <- ggplot(data = filtered_epi, mapping = aes(color = primaryDiagnosis, x = ageDeath, y = SST)) +
  geom_point() + 
  labs( title = "Chronological Age vs SST Proportions", 
        x = "Chronological Age", 
        y = "SST Proportion") +
  stat_smooth(method = "lm") + 
  stat_cor() 

chrono_age_vs_sst_proportions | epi_age_vs_sst_proportions
# VIP cells interesting

ggplot(data = filtered_epi, mapping = aes(x = model_residuals, color = primaryDiagnosis)) + 
  geom_histogram() + 
  labs( title = "Residuals of the Model", 
        x = "Residuals", 
        y = "Count")

# Now want to graph the residuals when binned by age
filtered_epi$predicted_ages = predicted_ages
filtered_epi$chrono_age_groups <- cut(x = filtered_epi$ageDeath, breaks = seq(10, 100, by = 10), labels = c("10-19", "20-29", "30-39", "40-49", "50-59", "60-69", "70-79", "80-89", "90+"))

filtered_epi$epi_age_groups <- cut(x = filtered_epi$predicted_ages, breaks = seq(10, 100, by = 10), labels = c("10-19", "20-29", "30-39", "40-49", "50-59", "60-69", "70-79", "80-89", "90+"))

ggplot(data = filtered_epi, mapping = aes(x = chrono_age_groups, y = model_residuals, color = primaryDiagnosis)) +
  geom_boxplot() + 
  labs( x = "Chronological Age Groups", y = "Residuals", title = "Residuals for Chronological Age Groups")

ggplot(data = filtered_epi, mapping = aes(x = epi_age_groups, y = model_residuals, color = primaryDiagnosis)) +
  geom_boxplot() + 
  labs( x = "Epigenetic Age Groups", y = "Residuals", title = "Residuals for Epigenetic Age Groups")


```

Import EpiDISH / CellDMC Required Packages
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("EpiDISH")
library(EpiDISH)
```

Loading example data / running workflow
```{r}
data(centEpiFibIC.m)
data(DummyBeta.m)
frac.m <- epidish(beta.m = DummyBeta.m, ref.m = centEpiFibIC.m, method = "RPC")
dim(frac.m$dataREF)
dim(frac.m$estF)
pheno.v <- rep(c(0, 1), each = 5)
dim(pheno.v)
celldmc.o <- CellDMC(DummyBeta.m, pheno.v, frac.m$estF)
head(celldmc.o$coe)
```

Loading our data
```{r}
# Load the probe data
beta_matrix <- read.csv('/Users/hugohudson/Downloads/GSE74193_GEO_procData.csv')
dim(beta_matrix)
# Load the patient information
patient_info <- read.csv('/Users/hugohudson/Desktop/GSE74193_series_matrix.csv')
# Extract phenotype information
patient_info <- data.frame(patient_info)
# Load the cell fractions 
hugo_epi_age = read.csv("/Users/hugohudson/Desktop/epigenetic_age_added.csv")
filtered_epi <- filter(hugo_epi_age,  Epigenetic.Age > 0)
dim(filtered_epi)
filtered_epi$individualID
# Check that the brnum samples are in alignment
brnums <- patient_info[43,2:ncol(patient_info)]
brnums <- substr(brnums, 50, nchar(brnums))
brnums_dan <- filtered_epi$individualID
filtered_brnums <- brnums_dan[brnums_dan %in% brnums]
filtered_data <- filter(filtered_epi, individualID %in% brnums)
# Create a dataframe with the brnums and cpg info
colnames(beta_matrix) <- brnums
sorted_df <- beta_matrix[, match(filtered_brnums, colnames(beta_matrix))]
# Now in same order as Dan's data, so can use filtered_epi
pheno <- filtered_epi$primaryDiagnosis
fracs <- filtered_epi[1:nrows(filtered_epi),32:ncol(filtered_epi)]
```

Running CellDMC workflow on our data
```{r}
celldmc.o <- CellDMC(beta_matrix, pheno, fracs)
head(celldmc.o$coe)
```

Load horvath clock
```{r}
# Load the package / functions
#install.packages("WGCNA")
#install.packages("sqldf")
#install.packages("BiocManager")
#BiocManager::install("GO.db")
# BiocManager::install("impute")
#install the Bioconductor installer
#install.packages("BiocInstaller",repos="http://www.bioconductor.org/packages/2.13/bioc")
#install “impute” from Bioconductor 
#biocLite("impute")
#BiocManager::install("preprocessCore")
library(WGCNA)
library(sqldf)
BiocManager::install("BMIQ")
library(BMIQ)
# Load the betas
#horvath_betas <- read.csv('/Users/hugohudson/Downloads/gb-2013-14-10-r115-S3.csv')
# Load the example data
library(RPMM)
```

Test horvath clock on horvath example data
```{r}
# Age transofrmation and probe annotation functions
trafo= function(x,adult.age=20) { x=(x+1)/(1+adult.age); y=ifelse(x<=1, log( x),x-1);y }
anti.trafo= function(x,adult.age=20) { ifelse(x<0, (1+adult.age)*exp(x)-1, (1+adult.age)*x+adult.age) }
probeAnnotation21kdatMethUsed=read.csv("/Users/hugohudson/Downloads/gb-2013-14-10-r115-S22.csv")
probeAnnotation27k=read.csv("/Users/hugohudson/Downloads/gb-2013-14-10-r115-S21.csv")
datClock=read.csv("/Users/hugohudson/Downloads/gb-2013-14-10-r115-S23.csv")

# Read in DNA methylation data
# For a small file, e.g. measured on the 27k platform you could just use read.csv. 
# But for large files, e.g. those measured on the 450K platform, I recommend you use read.csv.sql.
dat0=read.csv.sql("/Users/hugohudson/Downloads/gb-2013-14-10-r115-S26.csv") ;
nSamples=dim(dat0)[[2]]-1
nProbes= dim(dat0)[[1]]
# the following command may not be needed. But it is sometimes useful when you use read.csv.sql
dat0[,1]= gsub(x=dat0 [,1],pattern="\"",replacement="") 
#Create a log file which will be output into your directory
# The code looks a bit complicated because it serves to create a log file (for error checks etc).
# It will automatically create a log file.
file.remove("LogFile.txt")
file.create("LogFile.txt")
DoNotProceed=FALSE
cat(paste( "The methylation data set contains", nSamples, "samples (e.g. arrays) and ", nProbes, " probes."),file="LogFile.txt")
if (nSamples==0) {DoNotProceed=TRUE; cat(paste( "\n ERROR: There must be a data input error since there seem to be no samples.\n Make sure that you input a comma delimited file (.csv file)\n that can be read using the R command read.csv.sql . Samples correspond to columns in that file  ."), file="LogFile.txt",append=TRUE) } 
if (nProbes==0) {DoNotProceed=TRUE; cat(paste( "\n ERROR: There must be a data input error since there seem to be zero probes.\n Make sure that you input a comma delimited file (.csv file)\n that can be read using the R command read.csv.sql  CpGs correspond to rows.")   , file="LogFile.txt",append=TRUE) } 
if (  nSamples > nProbes  ) { cat(paste( "\n MAJOR WARNING: It worries me a lot that there are more samples than CpG probes.\n Make sure that probes correspond to rows and samples to columns.\n I wonder whether you want to first transpose the data and then resubmit them? In any event, I will proceed with the analysis."),file="LogFile.txt",append=TRUE) }
if (  is.numeric(dat0[,1]) ) { DoNotProceed=TRUE; cat(paste( "\n Error: The first column does not seem to contain probe identifiers (cg numbers from Illumina) since these entries are numeric values. Make sure that the first column of the file contains probe identifiers such as cg00000292. Instead it contains ", dat0[1:3,1]  ),file="LogFile.txt",append=TRUE)  } 
if (  !is.character(dat0[,1]) ) {  cat(paste( "\n Major Warning: The first column does not seem to contain probe identifiers (cg numbers from Illumina) since these entries are numeric values. Make sure that the first column of the file contains CpG probe identifiers such as cg00000292. Instead it contains ", dat0[1:3,1]  ),file="LogFile.txt",append=TRUE)  } 
datout=data.frame(Error=c("Input error. Please check the log file for details","Please read the instructions carefully."), Comment=c("", "email Steve Horvath."))
if ( ! DoNotProceed ) {
nonNumericColumn=rep(FALSE, dim(dat0)[[2]]-1)
for (i in 2:dim(dat0)[[2]] ){ nonNumericColumn[i-1]=! is.numeric(dat0[,i]) }
if (  sum(nonNumericColumn) >0 ) { cat(paste( "\n MAJOR WARNING: Possible input error. The following samples contain non-numeric beta values: ", colnames(dat0)[-1][ nonNumericColumn], "\n Hint: Maybe you use the wrong symbols for missing data. Make sure to code missing values as NA in the Excel file. To proceed, I will force the entries into numeric values but make sure this makes sense.\n" ),file="LogFile.txt",append=TRUE)  } 
XchromosomalCpGs=as.character(probeAnnotation27k$Name[probeAnnotation27k$Chr=="X"])
selectXchromosome=is.element(dat0[,1], XchromosomalCpGs )
selectXchromosome[is.na(selectXchromosome)]=FALSE
meanXchromosome=rep(NA, dim(dat0)[[2]]-1)
if (   sum(selectXchromosome) >=500 )  {
meanXchromosome= as.numeric(apply( as.matrix(dat0[selectXchromosome,-1]),2,mean,na.rm=TRUE)) }
if (  sum(is.na(meanXchromosome)) >0 ) { cat(paste( "\n \n Comment: There are lots of missing values for X chromosomal probes for some of the samples. This is not a problem when it comes to estimating age but I cannot predict the gender of these samples.\n " ),file="LogFile.txt",append=TRUE)  } 

match1=match(probeAnnotation21kdatMethUsed$Name , dat0[,1])
if  ( sum( is.na(match1))>0 ) { 
missingProbes= probeAnnotation21kdatMethUsed$Name[!is.element( probeAnnotation21kdatMethUsed$Name , dat0[,1])]    
DoNotProceed=TRUE; cat(paste( "\n \n Input error: You forgot to include the following ", length(missingProbes), " CpG probes (or probe names):\n ", paste( missingProbes, sep="",collapse=", ")),file="LogFile.txt",append=TRUE)  } 

#STEP 2: Restrict the data to 21k probes and ensure they are numeric
match1=match(probeAnnotation21kdatMethUsed$Name , dat0[,1])
if  ( sum( is.na(match1))>0 ) stop(paste(sum( is.na(match1)), "CpG probes cannot be matched"))
dat1= dat0[match1,]
asnumeric1=function(x) {as.numeric(as.character(x))}
dat1[,-1]=apply(as.matrix(dat1[,-1]),2,asnumeric1)

#STEP 3: Create the output file called datout
set.seed(1)
# Do you want to normalize the data (recommended)?
normalizeData=TRUE
source("/Users/hugohudson/Downloads/gb-2013-14-10-r115-S25.txt")

# STEP 4: Output the results 
if (  sum(  datout$Comment  != "" )   ==0 ) { cat(paste( "\n The individual samples appear to be fine. "),file="LogFile.txt",append=TRUE)  } 
if (  sum(  datout$Comment != "" )   >0 ) { cat(paste( "\n Warnings were generated for the following samples.\n", datout[,1][datout$Comment != ""], "\n Hint: Check the output file for more details."),file="LogFile.txt",append=TRUE)  } 
} 
# output the results into the directory
write.table(datout,"Output.csv", row.names=F, sep="," )

```

Use horvath to validate our previously calculated epigenetic ages
```{r}

```


Use horvath clock on CellDMC output
```{r}

```

